### PR TITLE
quick_union with weighted trees and path compression

### PR DESCRIPTION
**Pull request by MoustaphaSaad**

### CODE REVIEW

File: cpprelude/inc/cpprelude/dynamic_array.h [hunk 1]
```diff
@@ -4,13 +4,14 @@
...
 #include <initializer_list>
+#include <iterator>
 
 
 namespace cpprelude
 {
-	//configurations
-	 r32 grow_factor = 1.5f;
-	 usize starting_count = 64;
+//configurations
```
Comment (NoraRo) [inline cpprelude/inc/cpprelude/dynamic_array.h]: These lines differ from vs2013 and vs2015.

File: cpprelude/inc/cpprelude/quick_union.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+		explicit quick_union(usize count);
+
+		void
+		init(usize count);
+
+		void
+		connect(usize a, usize b);
+
+		bool
+		is_connected(usize a, usize b);
```
Comment (NoraRo) [inline cpprelude/inc/cpprelude/quick_union.h]: Should unify the names between quick_find and quick_union later on.

File: cpprelude/inc/cpprelude/quick_union.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+		usize
+		count() const;
```
Comment (NoraRo) [inline cpprelude/inc/cpprelude/quick_union.h]: Are these helper functions needed? or should we add it to quick_find too?

File: cpprelude/inc/cpprelude/quick_union.h [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+		usize
+		_root(usize a);
```
Comment (NoraRo) [inline cpprelude/inc/cpprelude/quick_union.h]: Same as count()

File: cpprelude/src/cpprelude/quick_union.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
+#include "cpprelude/quick_union.h"
+
```
Comment (NoraRo) [inline cpprelude/src/cpprelude/quick_union.cpp]: Should we unify the style of Cpp and .h? or all the code in .h?

File: cpprelude/inc/cpprelude/quick_union.h (hunk 2 above)
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/quick_union.h]: yes
we agreed on `connect`, `is_connected`

File: cpprelude/inc/cpprelude/quick_union.h (hunk 3 above)
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/quick_union.h]: i think it's helpful and it costs nothing so yes

File: cpprelude/inc/cpprelude/quick_union.h (hunk 4 above)
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/quick_union.h]: everything starting with _ you can treat as a private so this is my implementation this is not user API

File: cpprelude/src/cpprelude/quick_union.cpp (hunk 5 above)
Comment (MoustaphaSaad) [inline cpprelude/src/cpprelude/quick_union.cpp]: this is a library so put code in cpp if possible leave only the declarations in header files

**Pull request merged**